synapseclient/core/download/download_async.py

```diff
@@ -326,169 +266,286 @@ class _MultithreadedDownloader:
 
-                self._check_for_errors(request, completed_futures)
+        Arguments:
+            request: A DownloadRequest object specifying what Synapse file to download.
+            download_tasks: A set of asyncio tasks to download the file in chunks.
 
-        except BaseException:
-            # on any exception (e.g. KeyboardInterrupt), attempt to cancel any pending futures.
-            # if they are already running this won't have any effect though
-            for future in pending_futures:
-                future.cancel()
+        Returns:
+            None
+        """
+        cause = None
+        while download_tasks:
+            done_tasks, pending_tasks = await asyncio.wait(
+                download_tasks, return_when=asyncio.FIRST_COMPLETED
+            )
+            download_tasks = pending_tasks
+            for completed_task in done_tasks:
+                try:
+                    start_bytes, end_bytes = completed_task.result()
+                    del completed_task
+                    self._syn._parts_transfered_counter += 1
+
+                    # Garbage collect every 100 iterations
+                    if self._syn._parts_transfered_counter % 100 == 0:
+                        gc.collect()
+
+                    self._syn.logger.debug(
+                        f"Downloaded bytes {start_bytes}-{end_bytes} to {self._download_request.path}"
+                    )
+                except BaseException as ex:
+                    # on any exception (e.g. KeyboardInterrupt), attempt to cancel any pending futures.
+                    # if they are already running this won't have any effect though
+                    cause = ex
+
+                    with self._thread_lock:
+                        if self._aborted:
+                            # we've already aborted, no need to execute this logic again
+                            continue
+                        self._aborted = True
+
+                    try:
+                        self._syn.logger.exception(
+                            f"Failed downloading {self._download_request.object_id} to {self._download_request.path}"
+                        )
+                        os.remove(self._download_request.path)
+                    except FileNotFoundError:
+                        pass
+
+        self._close_progress_bar()
+        if cause:
+            raise cause
+
+    def _update_progress_bar(self, part_size: int) -> None:
+        """Update the progress bar with the given part size."""
+        if self._syn.silent or not self._progress_bar:
+            return
+        self._progress_bar.update(part_size)
+
+    def _create_progress_bar(self, file_size: int) -> None:
+        """Handle creating the progress bar.
 
-            try:
-                os.remove(request.path)
-            except FileNotFoundError:
-                pass
+        Arguments:
+            file_size: The size of the file to download
```

Comment:
To clarify: This `file_size` is the size of a file _or_ part, correct?

Reply:
This was the total size of the file. I've changed this implementation a bit after getting a context managed TQDM bar for the sync function. See https://github.com/Sage-Bionetworks/synapsePythonClient/pull/1111/files#diff-29894805101b443f78e9548b65256eaff42ea1fd15290de9ab09857dac4765a1 for the new implementation